homeassistant/components/stream/__init__.py

```diff
@@ -394,35 +402,40 @@ def _run_worker(self) -> None:
                 redact_credentials(str(self.source)),
             )
 
-        @callback
-        def worker_finished() -> None:
+        async def worker_finished() -> None:
             # The worker is no checking availability of the stream and can no longer track
             # availability so mark it as available, otherwise the frontend may not be able to
             # interact with the stream.
             if not self.available:
                 self._async_update_state(True)
+            # We can call remove_provider() sequentially as the wrapped _stop() function
+            # which blocks internally is only called when the last provider is removed.
             for provider in self.outputs().values():
-                self.remove_provider(provider)
+                await self.remove_provider(provider)
 
-        self.hass.loop.call_soon_threadsafe(worker_finished)
+        self.hass.add_job(worker_finished)
```

Comment:
If we know the target is a coroutine we can use `hass.create_task` from another thread.
```suggestion
        self.hass.create_task(worker_finished())
```

Reply:
Thanks, made all the changes suggested here.